### DISCUSSION TITLE
aws-cloudformation: Unable to addDependency to CfnStack

### DISCUSSION THREAD

**Discussion by ChrisLane** [aws/aws-cdk#24635] [General]
### Describe the bug

I am unable to add an S3 deployment as a dependency to a CfnStack as I would expect from [the documentation](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cloudformation.CfnStack.html#addwbrdependencytarget).

### Expected Behavior

I expect to be able to add an S3 deployment resource as a dependency of a CfnStack, resulting in that CfnStack not being deployed until the S3 deployment is complete.

### Current Behavior

I receive the following error when running `npm run cdk -- synth` in the example project attached.
```
TypeError: target.shouldSynthesize is not a function
    at CfnStack.addDependency (/home/chris/cloud/documents/cdk-deployment-dependency-issue-example/node_modules/aws-cdk-lib/core/lib/cfn-resource.js:1:4351)
    at new CdkDeploymentDependencyIssueExampleStack (/home/chris/cloud/documents/cdk-deployment-dependency-issue-example/lib/cdk-deployment-dependency-issue-example-stack.js:78:20)
    at Object.<anonymous> (/home/chris/cloud/documents/cdk-deployment-dependency-issue-example/bin/cdk-deployment-dependency-issue-example.js:7:1)
    at Module._compile (node:internal/modules/cjs/loader:1275:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1329:10)
    at Module.load (node:internal/modules/cjs/loader:1133:32)
    at Module._load (node:internal/modules/cjs/loader:972:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:83:12)
    at node:internal/main/run_main_module:23:47
```

### Reproduction Steps

1. Download and extract this example project: [cdk-deployment-dependency-issue-example.zip](https://github.com/aws/aws-cdk/files/10919610/cdk-deployment-dependency-issue-example.zip)
1. `npm install` in the project directory
2. `npm run cdk -- synth` in the project directory

### CDK CLI Version

2.67.0 (build b6f7f39)

### Node.js Version

v19.7.0

### OS

Arch Linux 6.2.2-zen1

### Language

Typescript

**Comment by pahud** [comment]
This seems to be an interesting use case.

I'll see what I can do with your given sample.

```js
 const thirdPartyBinariesDeployment = new s3deploy.BucketDeployment(
        this,
        "third-party-binaries-deploy",
        {
          sources: [s3deploy.Source.asset("./s3-assets/third-party-binaries")],
          destinationBucket: thirdPartyBinariesBucket
        });

    const parameters = {
      BuildkiteQueue: "default",
      VpcId: vpc.vpcId,
      Subnets: vpc.privateSubnets.map(subnet => subnet.subnetId),
      InstanceRoleName: "buildkite-role",
      ManagedPolicyARN: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(AWSManagedPolicies.AMAZON_S3_FULL_ACCESS).managedPolicyArn
      ],
      AssociatePublicIpAddress: false,
      InstanceType: "t3.medium",
      MinSize: 1,
      MaxSize: 1
    };

    const buildkiteStack = new cloudformation.CfnStack(this, "buildkite-agent", {
      templateUrl: "https://s3.amazonaws.com/buildkite-aws-stack/v5.17.0/aws-stack.yml",
      parameters: objectMap(parameters, v => v.toString())
    });
    buildkiteStack.addDependency(thirdPartyBinariesDeployment);
  }
  ```

**Comment by pahud** [comment]
I think you probably need to use

`buildkiteStack.node.addDependency(thirdPartyBinariesDeployment)` instead.

And if you run `cdk synth` you will see the dependency in its synthesized template.

<img width="880" alt="image" src="https://user-images.githubusercontent.com/278432/225362789-b6d1f26a-eef4-49e9-9f77-9b9ec0caa3b8.png">

I noticed you are using native javascript to author your CDK code. I will recommend using TypeScript instead that helps you better trouble-shooing from your IDE with  strong type check capabilities.

Moving this to discussion as this appears to be a general guidance issue.

**Reply by ChrisLane** [reply]
Synth works with that, thank you!
